internal/watcher/nginx_process_parser.go

```diff
@@ -0,0 +1,340 @@
+	"github.com/nginx/agent/v3/internal/uuid"
+)
+
+const (
+	withWithPrefix   = "with-"
+	withModuleSuffix = "module"
+)
+
+type (
+	NginxProcessParser struct {
+		executer exec.ExecInterface
+	}
+
+	Info struct {
+		ProcessID       int32
+		Version         string
+		Prefix          string
+		ConfPath        string
+		ConfigureArgs   map[string]interface{}
+		ExePath         string
+		LoadableModules []string
+		DynamicModules  []string
+	}
+)
+
+var (
+	_ processParser = (*NginxProcessParser)(nil)
+
+	versionRegex = regexp.MustCompile(`(?P<name>\S+)\/(?P<version>.*)`)
+)
+
+func NewNginxProcessParser() *NginxProcessParser {
+	return &NginxProcessParser{
+		executer: &exec.Exec{},
+	}
+}
+
+func (npp *NginxProcessParser) Parse(ctx context.Context, processes []*model.Process) []*v1.Instance {
+	processList := []*v1.Instance{}
+
+	nginxProcesses := filterNginxProcesses(processes)
+
+	for _, nginxProcess := range nginxProcesses {
+		_, ok := nginxProcesses[nginxProcess.Ppid]
+		if !ok {
+			nginxInfo, err := npp.getInfo(ctx, nginxProcess)
+			if err != nil {
+				slog.DebugContext(ctx, "Unable to get NGINX info", "pid", nginxProcess.Pid, "error", err)
+
+				continue
+			}
+
+			processList = append(processList, convertInfoToProcess(*nginxInfo))
+		}
+	}
+
+	return processList
+}
+
+func (npp *NginxProcessParser) getInfo(ctx context.Context, nginxProcess *model.Process) (*Info, error) {
+	exePath := nginxProcess.Exe
+
+	if exePath == "" {
+		exePath = npp.getExe(ctx)
+		if exePath == "" {
+			return nil, fmt.Errorf("unable to find NGINX exe for process %d", nginxProcess.Pid)
+		}
+	}
+
+	var nginxInfo *Info
+
+	outputBuffer, err := npp.executer.RunCmd(ctx, exePath, "-V")
+	if err != nil {
+		return nil, err
+	}
+
+	nginxInfo = parseNginxVersionCommandOutput(ctx, outputBuffer)
+
+	nginxInfo.ExePath = exePath
+	nginxInfo.ProcessID = nginxProcess.Pid
+
+	loadableModules := getLoadableModules(nginxInfo)
+	nginxInfo.LoadableModules = loadableModules
+
+	nginxInfo.DynamicModules = getDynamicModules(nginxInfo)
+
+	return nginxInfo, err
+}
+
+func (npp *NginxProcessParser) getExe(ctx context.Context) string {
+	exePath := ""
+
+	out, commandErr := npp.executer.RunCmd(ctx, "sh", "-c", "command -v nginx")
+	if commandErr == nil {
+		exePath = strings.TrimSuffix(out.String(), "\n")
+	}
+
+	if exePath == "" {
+		exePath = npp.defaultToNginxCommandForProcessPath()
+	}
+
+	if strings.Contains(exePath, "(deleted)") {
+		exePath = sanitizeExeDeletedPath(exePath)
+	}
+
+	return exePath
+}
+
+func (npp *NginxProcessParser) defaultToNginxCommandForProcessPath() string {
+	exePath, err := npp.executer.FindExecutable("nginx")
+	if err != nil {
+		return ""
+	}
+
+	return exePath
+}
+
+func sanitizeExeDeletedPath(exe string) string {
+	firstSpace := strings.Index(exe, "(deleted)")
+	if firstSpace != -1 {
+		return strings.TrimSpace(exe[0:firstSpace])
+	}
+
+	return strings.TrimSpace(exe)
+}
+
+func convertInfoToProcess(nginxInfo Info) *v1.Instance {
```

Comment:
nit: it's confusing that we call the outcome a "process" when the type is "instance"
```suggestion
func convertInfoToInstance(nginxInfo Info) *v1.Instance {
```